compiler/src/dotty/tools/dotc/transform/FirstTransform.scala

```diff
@@ -147,9 +147,11 @@ class FirstTransform extends MiniPhase with InfoTransformer { thisPhase =>
     if (tree.isType) {
       toTypeTree(tree)
     } else if (tree.name != nme.WILDCARD) {
-      // We want to constant-fold _some_ idents here - for instance, @switch needs to see literals in patterns.
-      // However, constant-foldable wildcards can occur in patterns, for instance as `case _: "a"`;
-      // we avoid constant-folding those as doing so would change the meaning of the pattern.
+      // We constant-fold all idents except wildcards.
+      // AFAIK, constant-foldable wildcard idents can only occur in patterns, for instance as `case _: "a"`.
+      // Constant-folding that would result in `case "a": "a"`, which changes the meaning of the pattern.
+      // Note that we _do_ want to constant-fold idents in patterns that _aren't_ wildcards -
+      // for example, @switch annotation needs to see inlined literals and not indirect references.
```

Comment:
@smarter I think this comment is at the precise level of detail that we want. I don't think going into the subtle differences of Bind/Ident nodes in patterns makes sense - if we did that consequently in the whole compiler, then more than half of the documentation would be permanently outdated.